sbnci/Modules/RecoEff/RecoEff_module.cc

```diff
@@ -0,0 +1,417 @@
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "canvas/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+//Sim Base
+#include "nusimdata/SimulationBase/MCParticle.h"
+#include "nusimdata/SimulationBase/MCTruth.h"
+
+//Reco Base
+#include "lardataobj/RecoBase/PFParticle.h"
+#include "lardataobj/RecoBase/Track.h"
+#include "lardataobj/RecoBase/Shower.h"
+#include "lardataobj/RecoBase/Hit.h"
+
+//Tools
+#include "canvas/Persistency/Common/FindManyP.h"
+#include "canvas/Persistency/Common/FindOneP.h"
+
+//LArSoft
+#include "larsim/Utils/TruthMatchUtils.h"
+#include "lardata/DetectorInfoServices/DetectorClocksService.h"
+
+//Root
+#include "art_root_io/TFileService.h"
+#include "TTree.h"
+
+class RecoEff;
+
+class RecoEff : public art::EDAnalyzer {
+public:
+  explicit RecoEff(fhicl::ParameterSet const& pset);
+
+  RecoEff(RecoEff const&) = delete;
+  RecoEff(RecoEff&&) = delete;
+  RecoEff& operator=(RecoEff const&) = delete;
+  RecoEff& operator=(RecoEff&&) = delete;
+
+  void analyze(art::Event const &e) override;
+
+private:
+
+  void ClearData();
+  void SetupHitsMap();
+  void ReconstructionProcessor(art::Event const &e);
+  void TruthProcessor(art::Event const &e);
+
+  std::vector<art::Ptr<recob::PFParticle> > GetPrimaryPFPs();
+  art::Ptr<recob::PFParticle> GetPFP(long unsigned int const &id);
+
+  float Purity(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID);
+  float Completeness(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID);
+  float TrackLength(art::Ptr<recob::Track> const &track);
+  float TrueTrackLength(art::Ptr<simb::MCParticle> const &particle);
+
+
+  art::Handle<std::vector<simb::MCTruth> > eHandleNeutrinos;
+  art::Handle<std::vector<recob::Hit> > eHandleHits;
+  art::Handle<std::vector<recob::Track> > eHandleTracks;
+  art::Handle<std::vector<recob::Shower> > eHandleShowers;
+  art::Handle<std::vector<recob::PFParticle> > eHandlePFPs;
+
+  detinfo::DetectorClocksData clockData = art::ServiceHandle<detinfo::DetectorClocksService>()->DataForJob();
+
+  std::string fNuGenModuleLabel, fLArGeantModuleLabel, fPFParticleModuleLabel,
+    fTrackModuleLabel, fShowerModuleLabel, fHitsModuleLabel;
+
+  TTree *fParticleTree;
+
+  int mc_trackID, mc_PDG;
+  float mc_x0, mc_y0, mc_z0, mc_xEnd, mc_yEnd, mc_zEnd, mc_pX0,
+    mc_pY0, mc_pZ0, mc_energy0, mc_momentum, mc_pXEnd, mc_pYEnd, mc_pZEnd,
+    mc_energyEnd, mc_mass, mc_theta_xy, mc_theta_yz, mc_theta_xz, mc_length, 
+    reco_nTracks, reco_nShowers, reco_shower_purity, reco_shower_completeness,
+    reco_track_purity, reco_track_completeness, reco_track_length, reco_shower_dEdx;
+  bool reco_isReconstructed;
+
+  std::map<int,int> n_showers_map, n_tracks_map;
+  std::map<int,float> shower_comp_map, shower_pur_map, track_comp_map,
+    track_pur_map, track_length_map, shower_dEdx_map;
+  std::map<int,int> hits_map;
```

Comment:
you use fBlah above.  What's the reason for defining these differently?

Reply:
The mc_ and reco_ are per particle not really global in the same sense? The maps could be changed.

Reply:
I get what you mean but, technically speaking, they are as global as the `fBlah` variables (neither are global but both exist for equally as long as each other).  It might help with readability if its pointed out that those variables are data members of the class.

Reply:
That's true, I'll update them!